File: number.go
//JavaScript中的number类型和Number对象

package ss

import (
	"bytes"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"math"
	"ss/collection"
	"strconv"
	"strings"
	"unicode"
)

// Number JavaScript Number类型
type Number struct {
	Data float64
}

var _ JavaScript = &Number{}
var _ JavaScriptBasic = &Number{}
var _ CodeCompiler = &Number{}

func init() {
	//注册一个全局对象Number
	o := NewDefaultObject("Number")
	o.AddProperty("isFinite", isFinite())
	o.AddProperty("isInteger", isInteger())
	o.AddProperty("isNaN", isNaN1())
	o.AddProperty("isSafeInteger", isSafeInteger())
	o.AddProperty("parseFloat", NewParseFloat())
	o.AddProperty("parseInt", NewParseInt())
	o.AddProperty("MAX_VALUE", GoToNumber(math.MaxFloat64))
	o.AddProperty("MIN_VALUE", GoToNumber(math.SmallestNonzeroFloat64))
	setGlobalObject("Number", o)
}

// GoToNumber 将go语言中的float64转换为js中的number
func GoToNumber[T float64 | float32 | int | int8 | int16 | int32 | int64 | uint8 | uint16 | uint32 | uint64 | string](d T) *Number {
	o := &Number{}
	var a any = d
	switch t := a.(type) {
	case float64:
		o.Data = t
	case int:
		o.Data = float64(t)
	case float32:
		o.Data = float64(t)
	case int8:
		o.Data = float64(t)
	case int16:
		o.Data = float64(t)
	case int32:
		o.Data = float64(t)
	case int64:
		o.Data = float64(t)
	case uint8:
		o.Data = float64(t)
	case uint16:
		o.Data = float64(t)
	case uint32:
		o.Data = float64(t)
	case uint64:
		o.Data = float64(t)
	case string:
		i, e := strconv.ParseFloat(t, 64)
		if e != nil {
			o.Data = math.NaN()
		} else {
			o.Data = i
		}
	default:
		panic("not support type ")
	}
	return o
}

// NewNaN 新建NaN
func NewNaN() *Number {
	return GoToNumber(math.NaN())
}

// IsNaN 判断js变量是否为NaN
func IsNaN(s JavaScript) bool {
	t := JsToType(s)
	if t.Type != NumberType {
		return false
	}
	return math.IsNaN(s.(*Number).Data)
}

// 将16进制表示的字符串转换为js中的number
func hexStrToNumber(hexStr string) JavaScript {
	cleanHexStr := strings.TrimPrefix(strings.ToLower(hexStr), "0x")
	i, e := strconv.ParseInt(cleanHexStr, 16, 64)
	if e != nil {
		return NewNaN()
	}
	return GoToNumber(float64(i))
}

// 将8进制表示的字符串转换为js中的number
func octalStrToNumber(octalStrToInt string) JavaScript {
	nStr := strings.TrimPrefix(strings.ToLower(octalStrToInt), "0o")
	i, e := strconv.ParseInt(nStr, 8, 64)
	if e != nil {
		return NewNaN()
	}
	return GoToNumber(float64(i))
}

// 将2进制表示的字符串转换为js中的number
func binaryStrToNumber(octalStrToInt string) JavaScript {
	nStr := strings.TrimPrefix(strings.ToLower(octalStrToInt), "0b")
	i, e := strconv.ParseInt(nStr, 2, 64)
	if e != nil {
		return NewNaN()
	}
	return GoToNumber(float64(i))
}

func floatToString(number float64, radix int) string {
	if radix < 2 || radix > 36 {
		panic(NewRangeError(fmt.Sprintf("radix out of range: %d", radix)))
	}
	// 对于十进制，使用 FormatFloat 并保留原始的浮点数精度。
	if radix == 10 {
		// -1 表示使用最短必要的小数位数，'f' 表示不使用科学计数法。
		return strconv.FormatFloat(number, 'f', -1, 64)
	}
	// 对于其他进制，仅处理整数部分。
	intPart := int64(math.Floor(math.Abs(number)))
	formatted := strconv.FormatInt(intPart, radix)
	if number < 0 {
		formatted = "-" + formatted
	}
	return formatted
}

func numberToStr(n *Number) JavaScript {
	return NewDefaultFunction("toString", func(args []JavaScript, ctx *Context) JavaScript {
		if math.IsInf(n.Data, 1) {
			return GoToString("Infinity")
		}
		if math.IsInf(n.Data, -1) {
			return GoToString("-Infinity")
		}
		if math.IsNaN(n.Data) {
			return GoToString("NaN")
		}
		if len(args) == 0 {
			return GoToString(n.ToString())
		}
		radix := 10
		if j, ok := args[0].(*Number); ok {
			radix = int(j.Data)
		}
		if radix == 10 {
			return GoToString(n.ToString())
		}
		return GoToString(floatToString(n.Data, radix))
	})
}

func numberToExponential(n *Number) JavaScript {
	return NewDefaultFunction("toExponential", func(args []JavaScript, ctx *Context) JavaScript {
		if math.IsInf(n.Data, 1) {
			return GoToString("Infinity")
		}
		if math.IsInf(n.Data, -1) {
			return GoToString("-Infinity")
		}
		if math.IsNaN(n.Data) {
			return GoToString("NaN")
		}
		d := 0
		if len(args) > 0 {
			if j, ok := args[0].(*Number); ok {
				d = int(j.Data)
			}
		}
		if d < 0 || d > 100 {
			panic(NewRangeError("toExponential() digits argument must be between 0 and 100"))
		}
		format := fmt.Sprintf("%%.%de", d)
		if len(args) == 0 {
			format = "%e"
		}
		formatted := fmt.Sprintf(format, n.Data)
		return GoToString(formatted)
	})
}

func isFinite() JavaScript {
	return NewDefaultFunction("isFinite", func(args []JavaScript, ctx *Context) JavaScript {
		if len(args) == 0 {
			return GoToBoolean(false)
		}
		if d, ok := args[0].(*Number); ok {
			if math.IsInf(d.Data, 1) || math.IsInf(d.Data, -1) || math.IsNaN(d.Data) {
				return GoToBoolean(false)
			}
			return GoToBoolean(true)
		} else {
			return GoToBoolean(false)
		}
	})
}

func isInteger() JavaScript {
	return NewDefaultFunction("isInteger", func(args []JavaScript, ctx *Context) JavaScript {
		if len(args) == 0 {
			return GoToBoolean(false)
		}
		if d, ok := args[0].(*Number); ok {
			if math.IsInf(d.Data, 1) || math.IsInf(d.Data, -1) || math.IsNaN(d.Data) {
				return GoToBoolean(false)
			}
			return GoToBoolean(d.Data == math.Floor(d.Data))
		} else {
			return GoToBoolean(false)
		}
	})
}

func isNaN1() JavaScript {
	return NewDefaultFunction("isNaN", func(args []JavaScript, ctx *Context) JavaScript {
		if len(args) == 0 {
			return GoToBoolean(false)
		}
		if d, ok := args[0].(*Number); ok {
			return GoToBoolean(math.IsNaN(d.Data))
		} else {
			return GoToBoolean(false)
		}
	})
}

func NewParseFloat() JavaScript {
	c := func(args []JavaScript, r *Context) JavaScript {
		if len(args) == 0 {
			return NewUndefined()
		}
		str := strings.TrimLeftFunc(JsToString(args[0]), unicode.IsSpace)
		return GoToNumber(getLeadingFloat(str))
	}
	return NewDefaultFunction("parseFloat", c)
}

func getLeadingFloat(s string) string {
	var result []rune
	for _, r := range s {
		if unicode.IsDigit(r) || '.' == r {
			result = append(result, r)
		} else {
			break
		}
	}
	return string(result)
}
func isSafeInteger() JavaScript {
	return NewDefaultFunction("isSafeInteger", func(args []JavaScript, ctx *Context) JavaScript {
		if len(args) == 0 {
			return GoToBoolean(false)
		}
		if d, ok := args[0].(*Number); ok {
			const MaxInt53 = 1<<53 - 1
			return GoToBoolean(d.Data == math.Floor(d.Data) && d.Data >= -MaxInt53 && d.Data <= MaxInt53)
		} else {
			return GoToBoolean(false)
		}
	})
}

func NewParseInt() JavaScript {
	c := func(args []JavaScript, r *Context) JavaScript {
		if len(args) == 0 {
			return NewUndefined()
		}
		base := 10
		if len(args) > 1 {
			if a, ok := args[1].(*Number); ok {
				base = int(a.Data)
			} else {
				return NewNaN()
			}
		}
		s := JsToString(args[0])
		if strings.HasPrefix(s, "0x") {
			if len(args) == 1 {
				base = 16
			}
			f, e := strconv.ParseInt(s[2:], base, 64)
			if e != nil {
				return NewNaN()
			}
			return GoToNumber(float64(f))
		}
		str := getLeadingDigits(strings.TrimLeftFunc(s, unicode.IsSpace), base)
		if str == "" {
			return NewNaN()
		}
		f, e := strconv.ParseInt(str, base, 64)
		if e != nil {
			return NewNaN()
		}
		return GoToNumber(float64(f))
	}
	return NewDefaultFunction("parseInt", c)
}

func getLeadingDigits(s string, base int) string {
	var result []rune
	for _, r := range s {
		if !unicode.IsDigit(r) && !('a' <= r && r <= 'f') && !('A' <= r && r <= 'F') {
			break
		}
		switch base {
		case 2:
			if r == '0' || r == '1' {
				result = append(result, r)
			} else {
				break
			}
		case 8:
			if '0' <= r && r <= '7' {
				result = append(result, r)
			} else {
				break
			}
		case 10:
			if unicode.IsDigit(r) {
				result = append(result, r)
			} else {
				break
			}
		case 16:
			if unicode.IsDigit(r) || ('a' <= r && r <= 'f') || ('A' <= r && r <= 'F') {
				result = append(result, r)
			} else {
				break
			}
		}
		if base != 2 && base != 8 && base != 10 && base != 16 {
			break
		}
	}
	return string(result)
}

func numberToFixed(n *Number) JavaScript {
	return NewDefaultFunction("toFixed", func(args []JavaScript, ctx *Context) JavaScript {
		d := 0
		if len(args) > 0 {
			if j, ok := args[0].(*Number); ok {
				d = int(j.Data)
			}
		}
		if d < 0 || d > 100 {
			panic(NewRangeError("toFixed() digits argument must be between 0 and 100"))
		}
		formatted := fmt.Sprintf("%.*f", d, n.Data)
		return GoToString(formatted)
	})
}

func numberToPrecision(n *Number) JavaScript {
	return NewDefaultFunction("toPrecision", func(args []JavaScript, ctx *Context) JavaScript {
		d := 0
		if len(args) > 0 {
			if j, ok := args[0].(*Number); ok {
				d = int(j.Data)
			}
		} else {
			return GoToString(n.ToString())
		}
		if d < 0 || d > 100 {
			panic(NewRangeError("toPrecision() digits argument must be between 0 and 100"))
		}
		// 特殊情况处理：0 和非数值
		if n.Data == 0 {
			return GoToString("0." + strings.Repeat("0", d-1))
		}
		order := int(math.Log10(math.Abs(n.Data)))
		var formatStr string
		if order >= d || order < -4 {
			// 使用科学记数法格式
			formatStr = fmt.Sprintf("%%.%de", d-1)
		} else {
			// 使用固定小数点格式
			decimalPlaces := d - order - 1
			if n.Data < 1 {
				decimalPlaces = d - order
			}
			if decimalPlaces < 0 {
				decimalPlaces = 0
			}
			formatStr = fmt.Sprintf("%%.%df", decimalPlaces)
		}
		return GoToString(fmt.Sprintf(formatStr, n.Data))
	})
}

/*****************JavaScript接口实现******************/

func (n *Number) GetName() string {
	return "number"
}

func (n *Number) GetProperty(name string) JavaScript {
	switch name {
	case "toFixed":
		return numberToFixed(n)
	case "toPrecision":
		return numberToPrecision(n)
	case "toString":
		return numberToStr(n)
	case "toExponential":
		return numberToExponential(n)
	default:
		return NewUndefined()
	}

}

func (n *Number) GetProperties() JavaScriptProperties {
	return collection.NewArrayMap[string, JavaScript]()
}

func (n *Number) ToString() string {
	if math.IsInf(n.Data, 1) {
		return "Infinity"
	}
	if math.IsInf(n.Data, -1) {
		return "-Infinity"
	}
	if math.IsNaN(n.Data) {
		return "NaN"
	}
	//打印
	var formatStr string
	intPartLength := int(math.Log10(math.Abs(n.Data))) + 1
	if math.Abs(n.Data) < 1e-6 {
		//当绝对值小于1e-6（不包括1e-6本身）时，也会使用科学计数法表示。
		formatStr = "%v"
	} else if n.Data > 0 {
		//对于正数
		if n.Data >= 1e+21 {
			//数字大于或等于1e+21时，JavaScript会自动使用科学计数法表示。
			formatStr = "%v"
		} else {
			//正常显示
			if intPartLength < 17 {
				x := getFractionalPartDigits(n.Data)
				if x > 0 {
					formatStr = fmt.Sprintf("%%.%df", x)
				} else {
					formatStr = "%" + strconv.Itoa(intPartLength) + ".0f"
				}
			} else {
				formatStr = "%" + strconv.Itoa(intPartLength) + ".0f"
			}
		}
	} else {
		//对于正数
		if n.Data <= -1e+21 {
			//小于或等于-1e+21的值，JavaScript会自动使用科学计数法表示。
			formatStr = "%v"
		} else {
			//正常显示
			if intPartLength < 17 {
				x := getFractionalPartDigits(n.Data)
				if x > 0 {
					formatStr = fmt.Sprintf("%%.%df", x)
				} else {
					formatStr = "%" + strconv.Itoa(intPartLength) + ".0f"
				}
			} else {
				formatStr = "%" + strconv.Itoa(intPartLength) + ".0f"
			}
		}
	}
	return fmt.Sprintf(formatStr, n.Data)
}

func getFractionalPartDigits(value float64) int {
	// 将float64转换为字符串
	strValue := strconv.FormatFloat(value, 'f', -1, 64)
	// 分割整数部分和小数部分
	parts := strings.Split(strValue, ".")
	// 如果没有小数部分，返回0
	if len(parts) == 1 {
		return 0
	}
	// 返回小数部分的长度
	return len(parts[1])
}

func (n *Number) UnmarshalBin(data []byte) (any, error) {
	var f float64
	str := string(data)
	switch str {
	case `"+Inf"`:
		f = math.Inf(1)
	case `"-Inf"`:
		f = math.Inf(-1)
	case `"NaN"`:
		f = math.NaN()
	default:
		// 对于正常的float64值，直接解析
		value, err := strconv.ParseFloat(str, 64)
		if err != nil {
			return nil, err
		}
		f = value
	}
	n.Data = f
	return nil, nil
}

func (n *Number) MarshalBin() ([]byte, error) {
	// 检查是否为正负无穷大或NaN
	var da []byte
	if math.IsInf(n.Data, 1) {
		da = []byte(`"+Inf"`)
	} else if math.IsInf(n.Data, -1) {
		da = []byte(`"-Inf"`)
	} else if math.IsNaN(n.Data) {
		da = []byte(`"NaN"`)
	} else {
		da = []byte(strconv.FormatFloat(n.Data, 'f', -1, 64))
	}
	// 对于正常的float64值，使用标准格式
	data := []byte{5}
	data = append(data, da...)
	return data, nil
}

func (n *Number) ToBool() bool {
	if IsNaN(n) {
		return false
	}
	return n.Data != 0
}

func (n *Number) ToNumber() float64 {
	return n.Data
}
func (n *Number) Print() string {
	if math.IsInf(n.Data, 1) {
		return YellowStyle + "Infinity" + EndStyle
	}
	if math.IsInf(n.Data, -1) {
		return YellowStyle + "-Infinity" + EndStyle
	}
	if math.IsNaN(n.Data) {
		return YellowStyle + "NaN" + EndStyle
	}
	return YellowStyle + n.ToString() + EndStyle
}

func (n *Number) Value() any {
	return n.Data
}

func (n *Number) UnmarshalJSON(data []byte) error {
	return json.Unmarshal(data, &n.Data)
}
func (n *Number) MarshalJSON() ([]byte, error) {
	if math.IsInf(n.Data, +1) || math.IsInf(n.Data, -1) || math.IsNaN(n.Data) {
		return []byte(`null`), nil
	}
	return json.Marshal(n.Data)
}

func (n *Number) Binary() []byte {
	var buf bytes.Buffer
	err := binary.Write(&buf, binary.LittleEndian, n.Data)
	if err != nil {
		panic(NewError("Error", err.Error()))
	}
	return buf.Bytes()
}

File: test/ss/datatype/number.js
import assert from "assert";

console.log("测试基本数据类型::number");

// 测试Number.isFinite()的各种情况
assert.equal(Number.isFinite(Infinity), false, "Infinity不是有限数");
assert.equal(Number.isFinite(-Infinity), false, "-Infinity不是有限数");
assert.equal(Number.isFinite(NaN), false, "NaN不是有限数");
assert.equal(Number.isFinite('0'), false, "'0'不是Number类型，应返回false");
assert.equal(Number.isFinite(null), false, "null不是Number类型，应返回false");
assert.equal(Number.isFinite(0), true, "0是有限数");
assert.equal(Number.isFinite(2e64), true, "2e64是有限数");

// 测试Number.isInteger()的各种情况
assert.equal(Number.isInteger(0), true, "0是整数");
assert.equal(Number.isInteger(1), true, "1是整数");
assert.equal(Number.isInteger(-100000), true, "-100000是整数");
assert.equal(Number.isInteger(0.1), false, "0.1不是整数");
assert.equal(Number.isInteger(Math.PI), false, "π不是整数");
assert.equal(Number.isInteger(Infinity), false, "Infinity不是整数");
assert.equal(Number.isInteger(-Infinity), false, "-Infinity不是整数");
assert.equal(Number.isInteger("10"), false, "字符串'10'不是整数");
assert.equal(Number.isInteger(null), false, "null不是整数");
assert.equal(Number.isInteger(undefined), false, "undefined不是整数");
assert.equal(Number.isInteger(NaN), false, "NaN不是整数");

// 测试Number.isNaN()的各种情况
assert.equal(Number.isNaN(NaN), true, "NaN应该返回true");
assert.equal(Number.isNaN(0 / 0), true, "0除以0应该返回true");
assert.equal(Number.isNaN(NaN), true, "Number.NaN应该返回true");
assert.equal(Number.isNaN(0), false, "0不是NaN");
assert.equal(Number.isNaN(null), false, "null不是NaN");
assert.equal(Number.isNaN(undefined), false, "undefined不是NaN");
assert.equal(Number.isNaN('NaN'), false, "'NaN'字符串不是NaN");
assert.equal(Number.isNaN(''), false, "空字符串不是NaN");
assert.equal(Number.isNaN(true), false, "布尔值true不是NaN");

// 正常值
assert.equal(Number.isSafeInteger(0), true, "0是安全整数");
assert.equal(Number.isSafeInteger(1), true, "1是安全整数");
assert.equal(Number.isSafeInteger(-1), true, "-1是安全整数");
assert.equal(Number.isSafeInteger(Math.pow(2, 53) - 1), true, "2^53 - 1是安全整数的上限");
assert.equal(Number.isSafeInteger(-(Math.pow(2, 53) - 1)), true, "-(2^53 - 1)是安全整数的下限");

// 边界值
assert.equal(Number.isSafeInteger(Math.pow(2, 53)), false, "2^53超出安全整数范围");
assert.equal(Number.isSafeInteger(-(Math.pow(2, 53))), false, "-2^53超出安全整数范围");

// 非法值
assert.equal(Number.isSafeInteger("1024"), false, "字符串不是安全整数");
assert.equal(Number.isSafeInteger(NaN), false, "NaN不是安全整数");
assert.equal(Number.isSafeInteger(Infinity), false, "Infinity不是安全整数");
assert.equal(Number.isSafeInteger(-Infinity), false, "-Infinity不是安全整数");
assert.equal(Number.isSafeInteger(3.14), false, "浮点数不是安全整数");

// 正常值
assert.equal(Number.parseFloat("3.14"), 3.14, "'3.14' 应解析为 3.14");
assert.equal(Number.parseFloat("10.00"), 10.00, "'10.00' 应解析为 10");

// 边界值和非法值
assert.equal(Number.parseFloat("34 45 66"), 34, "'34 45 66' 应解析为 34，因为解析停在第一个空格");
assert.equal(Number.parseFloat("   60 "), 60, "'   60 ' 应解析为 60，忽略前后空格");

// 特殊情况
assert.equal(Number.parseFloat(""), NaN, "空字符串应解析为 NaN");
assert.equal(Number.parseFloat("abc"), NaN, "'abc' 应解析为 NaN");
assert.equal(Number.parseFloat("NaN"), NaN, "'NaN' 字符串应解析为 NaN");
assert.equal(Number.parseFloat("123abc"), 123, "'123abc' 应解析为 123，忽略数字后的文本");

// 确保使用assert.strictEqual来比较NaN值
assert.equal(Number.isNaN(Number.parseFloat("abc")), true, "'abc' 应解析为 NaN");

// 正常值
assert.equal(Number.parseInt('0x10'), 16, "'0x10' 应该被解析为十六进制整数 16");
assert.equal(Number.parseInt('10'), 10, "'10' 应该被解析为整数 10");
assert.equal(Number.parseInt('010'), 10, "'010' 应该被解析为整数 10，忽略前导零");

// 非法值和边界值
assert.equal(Number.parseInt(''), NaN, "空字符串应该被解析为 NaN");
assert.equal(Number.parseInt('abc'), NaN, "'abc' 应该被解析为 NaN");
assert.equal(Number.parseInt('10abc'), 10, "'10abc' 应该被解析为整数 10，忽略后续非数字字符");

// 使用assert.equal检查NaN
assert.equal(Number.isNaN(Number.parseInt('abc')), true, "'abc' 解析为 NaN");

// 正常值
assert.equal((123.456).toFixed(2), '123.46', '两位小数的四舍五入处理失败');
assert.equal((0).toFixed(0), '0', '整数0转换失败');
assert.equal((-123.456).toFixed(3), '-123.456', '负数保留三位小数失败');

// 边界值
assert.equal((1.005).toFixed(2), '1.00', '边界值四舍五入失败');
assert.equal((Number.MIN_VALUE).toFixed(2), '0.00', '最小值转换失败');

// 正常值测试
assert.equal((123.456).toPrecision(4), '123.5', '四舍五入失败');
assert.equal((123).toPrecision(3), '123', '整数返回失败');
assert.equal((0.000123).toPrecision(2), '0.00012', '小数点位数错误');
assert.equal((100000000000000).toPrecision(3), '1.00e+14', '科学计数法错误');

// 边界值测试
assert.equal((0.000000000000001).toPrecision(2), '1.0e-15', '极小值转换失败');
assert.equal((100000000000000000000).toPrecision(2), '1.0e+20', '极大值转换失败');
assert.equal((1.0001).toPrecision(3), '1.00', '超过小数点位数的四舍五入错误');

// 正常值测试
assert.equal((123).toString(), '123', '整数转换失败');
assert.equal((123.456).toString(), '123.456', '小数转换失败');
assert.equal((0).toString(), '0', '零转换失败');
assert.equal((0.1 + 0.2).toString(), '0.30000000000000004', '小数运算结果转换失败');

// 边界值测试
assert.equal((Number.MAX_VALUE).toString(), Number.MAX_VALUE.toString(), '最大值转换失败');
assert.equal((Number.MIN_VALUE).toString(), Number.MIN_VALUE.toString(), '最小值转换失败');

// 非法值测试
assert.equal((NaN).toString(), 'NaN', 'NaN值转换失败');
assert.equal((Infinity).toString(), 'Infinity', '正无穷大值转换失败');
assert.equal((-Infinity).toString(), '-Infinity', '负无穷大值转换失败');

// 正常值测试
assert.equal((123).toString(2), '1111011', '二进制转换失败');
assert.equal((123).toString(8), '173', '八进制转换失败');
assert.equal((123).toString(16), '7b', '十六进制转换失败');
assert.equal((123).toString(10), '123', '十进制转换失败');

// 边界值测试
assert.equal((0).toString(2), '0', '零的二进制转换失败');
assert.equal((0).toString(8), '0', '零的八进制转换失败');
assert.equal((0).toString(16), '0', '零的十六进制转换失败');
assert.equal((0).toString(10), '0', '零的十进制转换失败');

assert.equal((255).toString(2), '11111111', '255的二进制转换失败');
assert.equal((255).toString(8), '377', '255的八进制转换失败');
assert.equal((255).toString(16), 'ff', '255的十六进制转换失败');
assert.equal((255).toString(10), '255', '255的十进制转换失败');

// 非法值测试
assert.equal((NaN).toString(2), 'NaN', 'NaN值转换失败');
assert.equal((Infinity).toString(2), 'Infinity', '正无穷大值转换失败');
assert.equal((-Infinity).toString(2), '-Infinity', '负无穷大值转换失败');

// 正常值测试
assert.equal((123.456).toExponential(2), '1.23e+02', '正常小数转换失败');
assert.equal((123456).toExponential(2), '1.23e+05', '正常整数转换失败');
assert.equal((0.000123).toExponential(2), '1.23e-04', '小于1的小数转换失败');
assert.equal((0.1 + 0.2).toExponential(2), '3.00e-01', '小数运算结果转换失败');

// 边界值测试
assert.equal((Number.MAX_VALUE).toExponential(2), '1.80e+308', '最大值转换失败');
assert.equal((Number.MIN_VALUE).toExponential(2), '4.94e-324', '最小值转换失败');

// 非法值测试
assert.equal((NaN).toExponential(2), 'NaN', 'NaN值转换失败');
assert.equal((Infinity).toExponential(2), 'Infinity', '正无穷大值转换失败');
assert.equal((-Infinity).toExponential(2), '-Infinity', '负无穷大值转换失败');